### PR TITLE
feat: Add `set_as_inventory_for` api function

### LINKED ISSUE #1: Feature request: `Form:set_as_inventory_for(player)`
I'd like to prefix this feature request with a statement that not only am I willing to make this PR for you, I actually already know how I'd do it. Just let me know.

---

## The problem

There's one feature that flow can't do right now, and that is the "player formspec," as mentioned by this api:

> * `set_inventory_formspec(formspec)`
>     * Redefine player's inventory form
>     * Should usually be called in `on_joinplayer`
>     * If `formspec` is `""`, the player's inventory is disabled.

## Alternative

I've actually experimented with this and found that if I wanted to _force_ flow to do this, it would work like this:

- Manually call `_render` (in the same way as your unit tests), using code from `show_form`
- Manually manage state, and more or less copy a cloned version of `on_fs_input`.

This second bullet point in here is not something I'd like to do - especially since if (or when) you update how `on_fs_input` works, that update will not cascade downstream to these copied functions.

## My proposal

- Split the private `show_form` into two functions
  - local function `prepare_form`, which does _everything_ that `show_form` does, but instead of calling `minetest.show_formspec`, it returns the formspec string.
  - `show_form`, which calls `prepare_form`, then calls `minetest.show_formspec` on the result.
- Add a new function `Form:set_as_inventory_for()`. I'm not confidant in the name, but the functionality is more important. It calls `prepare_form` and passes the result into `player:set_inventory_formspec()`. ~~An update function would not be needed, and the `Form:update` * functions would do nothing.~~ _This is incorrect. See below for solution_

Example of use:

```lua
local stupid_simple_inv = flow.make_gui(function (p, c)
	return gui.List{
		inventory_location = "current_player",
		list_name = "main",
		w = 8,
		h = 4,
	}
end)
minetest.register_on_joinplayer(function(player)
	stupid_simple_inv:set_as_inventory_for(player)
end)
```

### Problem 1 with this proposal

A new special case will have to be added for `update_form`. We don't want it to call `show_form` for an inventory formspec.

#### Potential solution for for problem 1

- Add a flag to the context that is checked for in `update_form` that will tell it to _not_ call update form, and instead re-set the inventory formspec. We may want to abstract the `Form:set_as_inventory_for(player)` function for this to work.

This would actually mean that the `update` * functions would work as well.

### Problem 2 with this proposal

Players can only have one open formspec at a time. This would occupy that slot.

#### Potential solution for problem 2

Inventory Formspecs are always named `""`. This special case can be checked for and a different form info variable used in the multiple locations involved. An abstraction like that used in sfinv to set or unset form info (in sfinv's case it was ctx only) would be wise.

### PR DESCRIPTION
**Pull request by Lazerbeak12345**
Adds the `Form:set_as_inventory_for(player, [ctx])` api.

Also includes

- `Form:unset_as_inventory_for(player)`
- Extends private `show_form` to account for inv edge cases
- ~~Extends `Form:update` to account for inv edge cases~~
- ~~Extends `Form:update_where` to account for inv edge cases~~
- Extends private `on_fs_input` to account for inv edge cases
- Adds abstraction functions
- Add feature to `on_leaveplayer` to delete the inv formspec as well
- Add a few unit tests for related features

see #1

### CODE REVIEW

**Comment by Lazerbeak12345** [comment]
Doesn't include documentation right now

File: init.lua [hunk 1]
```diff
@@ -799,6 +820,11 @@ function Form:close_hud(player)
     hud_fs.close_hud(player, self)
 end
 
+function Form:unset_as_inventory_for(player)
+    open_inv_formspecs[player:get_player_name()] = nil
+    player:set_inventory_formspec("") -- TODO should this restore what it was before? This disables inventory as-is until set by something else.
```
Comment (Lazerbeak12345) [inline init.lua]: I'm not sure what you'd like this to do here.

File: init.lua [hunk 2]
```diff
@@ -811,18 +837,21 @@ local function update_form(self, player, form_info)
...
-            local player = minetest.get_player_by_name(name)
-            if player and func(player, form_info.ctx) then
-                update_form(self, player, form_info)
+    for _ , open_formspec_table in ipairs({open_formspecs, open_inv_formspecs}) do
+        for name, form_info in pairs(open_formspec_table) do
+            if form_info.self == self then
+                local player = minetest.get_player_by_name(name)
+                if player and func(player, form_info.ctx) then
+                    update_form(self, player, form_info)
+                end
```
Comment (luk3yx) [inline init.lua]: I don't think it's worth making `form:update` and `form:update_for` work with inventory formspecs, they don't work with HUDs at the moment and it should be possible to just call `set_as_inventory_for` again.

File: init.lua [hunk 3]
```diff
@@ -878,6 +918,7 @@ end
 
 local function on_leaveplayer(player)
     open_formspecs[player:get_player_name()] = nil
+    open_inv_formspecs[player:get_player_name()] = nil
 end
```
Comment (luk3yx) [inline init.lua]:
```suggestion
local function on_leaveplayer(player)
    local name = player:get_player_name()
    open_formspecs[name] = nil
    open_inv_formspecs[name] = nil
end
```

File: init.lua (hunk 1 above)
Comment (luk3yx) [reply init.lua]: I think this is okay, I'd be happy with it not changing the inventory formspec at all here but maybe it's a good idea to set it to `""` prevent confusion.

Maybe this function should do a check like `form:close` does so that you can't call unset_as_inventory_for with the wrong form object?

File: init.lua [hunk 4]
```diff
@@ -835,7 +864,14 @@ end
...
     local name = player:get_player_name()
-    local form_info = open_formspecs[name]
+    local inv = false
+    local form_info
+    if formname ~= "" then
+        form_info = open_formspecs[name]
+    else
+        form_info = open_inv_formspecs[name]
+        inv = true
+    end
```
Comment (luk3yx) [inline init.lua]:
```suggestion
    local form_infos = formname == "" and open_inv_formspecs or open_formspecs
    local form_info = form_infos[name]
```

File: init.lua [hunk 5]
```diff
@@ -866,9 +902,13 @@ local function on_fs_input(player, formname, fields)
         end
     end
 
-    if open_formspecs[name] ~= form_info then return true end
+    if inv then
+        if open_inv_formspecs[name] ~= form_info then return true end
+    else
+        if open_formspecs[name] ~= form_info then return true end
+    end
```
Comment (luk3yx) [inline init.lua]:
```suggestion
    if form_infos[name] ~= form_info then return true end
```

File: init.lua [hunk 6]
```diff
@@ -786,6 +803,10 @@ function Form:show_hud(player, ctx)
     hud_fs.show_hud(player, self, tree)
 end
 
+function Form:set_as_inventory_for(player, ctx)
+    set_inv_form(self, player, ctx or {})
+end
```
Comment (luk3yx) [inline init.lua:813]: I wonder whether it would be better to merge `set_inv_form` into this function (since it isn't used anywhere else) and make this function preserve the previous `ctx` (if the form is already set as the inventory and if the `ctx` argument isn't present) and `auto_name_id` (even if a different form was previously set).

File: init.lua [hunk 7]
```diff
@@ -866,9 +902,13 @@ local function on_fs_input(player, formname, fields)
...
 
-    if open_formspecs[name] ~= form_info then return true end
+    if inv then
+        if open_inv_formspecs[name] ~= form_info then return true end
+    else
+        if open_formspecs[name] ~= form_info then return true end
+    end
 
-    if fields.quit then
+    if fields.quit and not inv then
```
Comment (luk3yx) [inline init.lua]:
```suggestion
    if formname == "" then
        -- Special case for inventory forms
        if redraw_fs then
            form_info.self:set_as_inventory_for(player)
        end
    elseif fields.quit then
```

File: init.lua [hunk 8]
```diff
@@ -757,13 +756,31 @@ local function show_form(self, player, formname, ctx, auto_name_id)
     -- local t3 = DEBUG_MODE and minetest.get_us_time()
 
     form_info.formname = formname
-    open_formspecs[name] = form_info
     -- if DEBUG_MODE then
     --     print(t3 - t, t2 - t, t3 - t2)
     -- end
+    return  fs, form_info
```
Comment (luk3yx) [inline init.lua]:
```suggestion
    return fs, form_info
```

**Review by luk3yx — COMMENTED** [review]
Thank you for your PR! I've made some suggestions to make the code slightly more compact and to hopefully fix a bug (`update_form` will call `show_formspec` and not change the inventory formspec).

File: init.lua (hunk 1 above)
Comment (Lazerbeak12345) [reply init.lua]: Here's a test of what that would look like:

```lua
function Form:unset_as_inventory_for(player)
    local name = player:get_player_name()
    local form_info = open_inv_formspecs[name]
    if form_info and form_info.self == self then
        open_inv_formspecs[name] = nil
        player:set_inventory_formspec("")
    end
end
```

My concern is that it feels like since it's more or less the same as `Form:close`. Perhaps it'd be better to merge them? Or the duplication doesn't matter. Idk.

File: init.lua (hunk 2 above)
Comment (Lazerbeak12345) [reply init.lua]: Good point. There's actually a bug I missed here where it will call `minetest.show_formspec` on the updated form. Not sure if that's how it's supposed to be done.

Comment (Lazerbeak12345) [reply init.lua]: fixed

File: init.lua (hunk 6 above)
Comment (Lazerbeak12345) [reply init.lua:813]: done

**Comment by Lazerbeak12345** [comment]
Updated readme to reflect changes in `update` support.

File: init.lua [hunk 9]
```diff
@@ -786,6 +793,16 @@ function Form:show_hud(player, ctx)
     hud_fs.show_hud(player, self, tree)
 end
 
+local open_inv_formspecs = {}
+function Form:set_as_inventory_for(player, ctx)
+    local name = player:get_player_name()
+    -- Formname of "" is inventory
+    local fs, form_info = prepare_form(self, player, "", ctx or {})
```
Comment (luk3yx) [inline init.lua]: Maybe this function should preserve `ctx` if the same form is already set as the inventory?

```suggestion
    local old_form_info = open_inv_formspecs[name]
    if not ctx and old_form_info and old_form_info.self == self then
        ctx = old_form_info.ctx
    end

    -- Formname of "" is inventory
    local fs, form_info = prepare_form(self, player, "", ctx or {},
        old_form_info and old_form_info.auto_name_id)
```

`auto_name_id` stores the next number to use for automatic button naming to work around race conditions that would happen if button names were reused.

I may move the `auto_name_id < 1e6` check from `update_form` into `_render` later so it doesn't have to be duplicated here.

File: init.lua (hunk 1 above)
Comment (luk3yx) [reply init.lua]: The duplication here is okay, I think that trying to remove it would just make things more complicated.

Comment (Lazerbeak12345) [reply init.lua]: resolved in 0789c66

**Review by luk3yx — APPROVED** [review]
I'll try and test this PR and merge it soon

**Comment by Lazerbeak12345** [comment]
Would you like me to add the documentation to the readme?

**Comment by luk3yx** [comment]
> Would you like me to add the documentation to the readme?

Sure, in the "experimental features" section.

**Comment by luk3yx** [comment]
I've tested this PR and it works okay, I fixed a bug with remembering the values of scrollbars and checkboxes when the inventory is closed.

File: README.md [hunk 10]
```diff
@@ -399,3 +399,34 @@ gui.Stack{
 ```
 
 ![Screenshot](https://user-images.githubusercontent.com/3182651/215946217-3705dbd1-4ec8-4aed-a9eb-381fecb2d8f2.png)
+
+### Set form as player inventory
+
+A form can be set as the player inventory. Flow internally generates the
+formspec and passes it to `player:set_inventory_formspec()`.
```
Comment (luk3yx) [inline README.md]: Perhaps the incompatibility with other inventory mods (like sfinv) should be mentioned?

Comment (Lazerbeak12345) [reply README.md]: Sure. Something like "This will override your inventory mod?"

File: README.md [hunk 11]
```diff
@@ -399,3 +399,34 @@ gui.Stack{
...
+    example_inventory:set_as_inventory_for(player)
+end)
+```
+
+Like with the `show_hud` function, `update*` functions don't do anything, so to
+update it, call `set_as_inventory_for` again with the new context.
+
+```lua
+example_inventory:set_as_inventory_for(player, new_context)
+```
```
Comment (luk3yx) [inline README.md:425]: This should mention that calling `example_inventory:set_as_inventory_for(player)` will reuse the existing context

File: README.md (hunk 10 above)
Comment (luk3yx) [reply README.md]: Maybe "This will completely replace your inventory and isn't compatible with inventory mods like sfinv."

Comment (Lazerbeak12345) [reply README.md]: Oh, right. That's even better.

**Comment by Lazerbeak12345** [comment]
nice. I was just going to do the same and rebase or something.

**Comment by luk3yx** [comment]
Thanks again for your PR, good luck with your inventory mod!

**Comment by Lazerbeak12345** [comment]
Yeah, thanks for being so responsive! I'll watch for a new release on contentdb, and I'll likely stick around for #3 as well, even though at this point the API is enough for my needs.
